File: javascript/euler36.js
/* 

  Euler #36 in JavaScript.

  Problem 36:
  """
  The decimal number, 585 = 1001001001_(2) (binary), is palindromic 
  in both bases.
  
  Find the sum of all numbers, less than one million, which are palindromic 
  in base 10 and base 2.

  (Please note that the palindromic number, in either base, may not 
   include leading zeros.)
  """

  This JavaScript program was created by Hakan Kjellerstrand, hakank@gmail.com
  See also my JavaScript page: http://www.hakank.org/javascript_progs/

*/

'use strict';
const {range2,sum2,palindromic_list,dec2base,timing2} = require('./js_utils.js');

// 129ms
const euler36a = function() {
    let s = 0;
    for(let n = 1; n <= 999999; n++) {
        if (palindromic_list(n.toString().split("")) &&
            palindromic_list(dec2base(n,2))) {
            s += n;
        }
    }
    return s;
}

// Slower: 280ms
const euler36b = function() {
    return range2(1,999999)
        .filter(n=>palindromic_list(n.toString().split("")))
        .filter(n=>palindromic_list(dec2base(n,2)))
        .sum2();
}


timing2(euler36a); // 129ms
// timing2(euler36b); // 280ms

File: julia/euler36.jl
#=

  Euler #36 in Julia.

  Problem 36:
  """
  The decimal number, 585 = 1001001001_(2) (binary), is palindromic
  in both bases.

  Find the sum of all numbers, less than one million, which are palindromic
  in base 10 and base 2.

  (Please note that the palindromic number, in either base, may not
   include leading zeros.)
  """

  This Julia program was created by Hakan Kjellerstrand, hakank@gmail.com
  See also my Julia page: http://www.hakank.org/julia/

=#

include("Euler.jl")

# 0.06355465s
function euler36a()
    s = 0
    for n in 1:999999
        if palindromic_number(n) &&
            palindromic_list(digits(n,base=2))
            # palindromic_list(string(n,base=2))
            s += n;
        end
    end
    return s
end

# 0.19043701s
function euler36b()
    return ((1:999999)
        .|>n->n*palindromic_number(n)
        .|>n->n*palindromic_list(digits(n,base=2))
        )|>sum
end

run_euler(euler36a)
# run_euler(euler36b)
